tools/wds2idx.py

```diff
@@ -0,0 +1,183 @@
+            stderr=subprocess.PIPE,
+        )
+
+        tar_blocks = tar_blocks_proc.communicate()[0].split(b"\n")
+        tar_types_sizes = tar_types_sizes_proc.communicate()[0].split(b"\n")
+
+        last_blocks_line = None
+        for blocks_line in reversed(tar_blocks):
+            if not not blocks_line:
+                last_blocks_line = blocks_line
+                break
+
+        total_size = (
+            int(last_blocks_line[last_blocks_line.find(b"block") + 6 : last_blocks_line.find(b":")])
+            * 512
+        )
+
+        tar_data = zip(tar_blocks, tar_types_sizes)
+        tar_data = filter(lambda line: not not line[0] and not not line[1], tar_data)
+
+        # Aggregate extensions in samples
+        data = []
+        last_basename = None
+        for blocks_line, types_sizes_line in tar_data:
+            if counter % report_step == 0:
+                cur_time = time.time()
+                print(f"time: {cur_time - pre_time:.2f} count: {counter} stage: collect")
+            counter += 1
+
+            offset = int(blocks_line[blocks_line.find(b"block") + 6 : blocks_line.find(b":")]) * 512
+            name = str(blocks_line[blocks_line.find(b":") + 2 :], "ascii")
+            entry_type = types_sizes_line[0:1]
+
+            if entry_type != b"-" or name.startswith("."):
+                continue
+
+            # Extracting size
+            size = types_sizes_line[: -len(name)]
+            size = size[: size.rfind(b"-") - 8]  # "... <size> 20yy-mm-...."
+            size = int(size[size.rfind(b" ") :])
+
+            basename, extension = IndexCreator.split_name(name)
+
+            if last_basename != basename:
+                data.append((offset, [(extension, size)]))
+                last_basename = basename
+            else:
+                data[-1][1].append((extension, size))
+
+        if not data:
+            raise ValueError("Webdataset Tar File empty")
+
+        # Then construct the index file out of it
+        self.fidx.write(f"{total_size} {len(data)}\n")
+        for offset, extensions_sizes in data:
+            if counter % report_step == 0:
+                cur_time = time.time()
+                print(f"time: {cur_time - pre_time:.2f} count: {counter} stage: index")
+            self.fidx.write(
+                f"""{offset} {' '.join(
+                map(lambda ext_size: str(ext_size[0]) + ' ' + str(ext_size[1]), extensions_sizes)
+            )}\n"""
+            )
+            counter += 1
+
+        cur_time = time.time()
+        print(f"time: {cur_time - pre_time:.2f} count: {counter} stage: done")
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
+        description="Create an index file from .tar file",
+    )
+    parser.add_argument("archive", help="path to .tar file.")
+    parser.add_argument("index", help="path to index file.")
+    args = parser.parse_args()
```

Comment:
Please add some documentation to this script, i.e. how to use it and what is it used for.

Reply:
alright, just that I will be adding the documentation while the review goes along